### PR TITLE
Redshift to gcs

### PR DESCRIPTION
**Pull request by hhasija**

### CODE REVIEW

File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+export GCS_STAGING_LOCATION=<gcs-staging-bucket-folder> 
+export JARS=gs://<cloud-storage-bucket-name>/spark-redshift_<version>.jar
+
+bin/start.sh \
+-- --template REDSHIFTTOGCS \
+--templateProperty project.id=<gcp-project-id> \
+--templateProperty redshift.gcs.input.url=<jdbc:redshift://host-name:port-number/> \
+--templateProperty redshift.gcs.input.table=<Redshift-table-name> \
+--templateProperty redshift.gcs.temp.dir=<AWS-temp-directory> \
+--templateProperty redshift.gcs.iam.role=<Redshift-S3-IAM-role> \
```
[the remark this answers is not in the document]
Comment (shashank-google) [reply java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md:17]: also missing `.input` in the naming pattern

File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+--templateProperty redshift.gcs.access.key=<Access-key> \
+--templateProperty redshift.gcs.secret.key=<Secret-key> \
+--templateProperty redshift.gcs.file.format=<Output-File-Format> \
+--templateProperty redshift.gcs.file.location=<Output-GCS-location> \
+--templateProperty redshift.gcs.output.mode=<Output-GCS-Save-mode>
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md:22]: missing `.output` in the naming pattern

File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/RedshiftToGCS.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+    DataFrameWriter<Row> writer = inputData.write().format(fileFormat);
+    writer.mode(outputMode).save(fileLocation);
+
+    LOGGER.info("RedshiftToGcs job completed.");
+    spark.stop();
+  }
+
+  void validateInput() {
+    if (StringUtils.isAllBlank(inputUrl)
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/redshift/RedshiftToGCS.java:84]: Please use jakarta validation framework for input validation.
Refer to JdbcToGCSConfig for examples.

**Review by shashank-google — CHANGES_REQUESTED** [review]

File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
+## 1. RedShift To GCS
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md:1]: Missing changes in
/README.md
/java/README.md

**Review by shashank-google — CHANGES_REQUESTED** [review]

File: java/pom.xml [hunk 5]
```diff
@@ -142,6 +142,33 @@
       <artifactId>google-cloud-spanner-jdbc</artifactId>
       <version>2.4.2</version>
     </dependency>
+    <!-- https://mvnrepository.com/artifact/com.databricks/spark-redshift -->
+    <dependency>
+      <groupId>com.databricks</groupId>
+      <artifactId>spark-redshift_2.11</artifactId>
```
[the remark this answers is not in the document]
Comment (hhasija) [reply java/pom.xml]: Removing these dependencies from pom file, and adding them as export jar options.

File: java/pom.xml [hunk 6]
```diff
@@ -142,6 +142,33 @@
...
+      <version>3.0.0-preview1</version>
```
[the remark this answers is not in the document]
Comment (hhasija) [reply java/pom.xml]: Removing these dependencies from pom file, and adding them as export jar options.

File: java/pom.xml [hunk 7]
```diff
@@ -142,6 +142,33 @@
...
+    <!-- https://mvnrepository.com/artifact/com.amazon.redshift/redshift-jdbc42 -->
+    <dependency>
+      <groupId>com.amazon.redshift</groupId>
+      <artifactId>redshift-jdbc42</artifactId>
+      <version>2.1.0.9</version>
+    </dependency>
+    <!-- https://mvnrepository.com/artifact/org.apache.spark/spark-avro -->
+    <dependency>
+      <groupId>org.apache.spark</groupId>
+      <artifactId>spark-avro_2.13</artifactId>
```
[the remark this answers is not in the document]
Comment (hhasija) [reply java/pom.xml]: Removing these dependencies from pom file, and adding them as export jar options.

File: java/pom.xml [hunk 8]
```diff
@@ -142,6 +142,33 @@
...
+      <version>3.2.0</version>
+    </dependency>
+    <!-- https://mvnrepository.com/artifact/com.eclipsesource.minimal-json/minimal-json -->
+    <dependency>
+      <groupId>com.eclipsesource.minimal-json</groupId>
+      <artifactId>minimal-json</artifactId>
```
[the remark this answers is not in the document]
Comment (hhasija) [reply java/pom.xml]: Removing these dependencies from pom file, and adding them as export jar options.

File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+export SUBNET=<subnet>   
+export GCS_STAGING_LOCATION=<gcs-staging-bucket-folder> 
+export JARS=gs://<cloud-storage-bucket-name>/spark-redshift_<version>.jar
+
+bin/start.sh \
+-- --template REDSHIFTTOGCS \
+--templateProperty project.id=<gcp-project-id> \
+--templateProperty redshift.gcs.input.url=<jdbc:redshift://host-name:port-number/> \
+--templateProperty redshift.gcs.input.table=<Redshift-table-name> \
+--templateProperty redshift.gcs.temp.dir=<AWS-temp-directory> \
```
[the remark this answers is not in the document]
Comment (hhasija) [reply java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md]: Done

File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+--templateProperty redshift.gcs.iam.role=<Redshift-S3-IAM-role> \
+--templateProperty redshift.gcs.access.key=<Access-key> \
```
[the remark this answers is not in the document]
Comment (hhasija) [reply java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md]: Done

File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+--templateProperty redshift.gcs.secret.key=<Secret-key> \
```
[the remark this answers is not in the document]
Comment (hhasija) [reply java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md]: Done

File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+--templateProperty redshift.gcs.file.format=<Output-File-Format> \
+--templateProperty redshift.gcs.file.location=<Output-GCS-location> \
```
[the remark this answers is not in the document]
Comment (hhasija) [reply java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md]: Done

File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md (hunk 2 above)
Comment (hhasija) [reply java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md:22]: Done

File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/RedshiftToGCS.java (hunk 3 above)
Comment (hhasija) [reply java/src/main/java/com/google/cloud/dataproc/templates/redshift/RedshiftToGCS.java]: Done

File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+export REGION=<region>  
+export SUBNET=<subnet>   
+export GCS_STAGING_LOCATION=<gcs-staging-bucket-folder> 
+export JARS=gs://<cloud-storage-bucket-name>/spark-redshift_<version>.jar
+
+bin/start.sh \
+-- --template REDSHIFTTOGCS \
+--templateProperty project.id=<gcp-project-id> \
+--templateProperty redshift.gcs.input.url=<jdbc:redshift://host-name:port-number/> \
+--templateProperty redshift.gcs.input.table=<Redshift-table-name> \
```
[the remark this answers is not in the document]
Comment (hhasija) [reply java/src/main/java/com/google/cloud/dataproc/templates/redshift/README.md]: Done

File: java/src/main/java/com/google/cloud/dataproc/templates/databases/RedshiftToGCS.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    SparkSession spark =
+        SparkSession.builder()
+            .appName("Spark Template RedshiftToGCS ")
+            .enableHiveSupport()
+            .getOrCreate();
+
+    LOGGER.info("RedshiftToGcs job started.");
+
+    spark.sparkContext().hadoopConfiguration().set("fs.s3a.access.key", config.getAWSAccessKey());
+    spark.sparkContext().hadoopConfiguration().set("fs.s3a.secret.key", config.getAWSSecretKey());
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/databases/RedshiftToGCS.java:55]: duplicate code on LN 54 and 55

File: java/src/main/java/com/google/cloud/dataproc/templates/databases/RedshiftToGCSConfig.java [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+import jakarta.validation.constraints.Pattern;
+import java.util.Properties;
+
+public class RedshiftToGCSConfig {
+
+  static final ObjectMapper mapper =
+      new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+
+  @JsonProperty(value = REDSHIFT_GCS_OUTPUT_FILE_LOCATION)
+  @Pattern(regexp = "gs://(.*?)/(.*)")
```
Comment (shashank-google) [inline java/src/main/java/com/google/cloud/dataproc/templates/databases/RedshiftToGCSConfig.java:33]: great thanks for adding it

**Review by shashank-google — APPROVED** [review]
There is a duplicate line of code, see my comment.
Otherwise looks great !!

File: java/src/main/java/com/google/cloud/dataproc/templates/databases/RedshiftToGCS.java (hunk 14 above)
Comment (shashank-google) [reply java/src/main/java/com/google/cloud/dataproc/templates/databases/RedshiftToGCS.java:55]: oh it is not duplicate

**Pull request merged**